Adding labels to all job creation operation options.

amanda-tarafa opened the pull request:
A couple of notes for this PR:

- Labels have not been added as filters for the list jobs operation.
- The list jobs operation only populates job#configuration (and hence job#configuration.labels) if projection is set to full on the request parameters.
- We haven't yet surfaced projection as a parameter for list jobs operation, so we cannot even use the labels to filter the jobs client side.
- I didn't write snippets for labels because as long as we cannot filter there's no much use we can show, snippets would be only of setting labels.
- Is it OK if I surface projection, seems simple enough, anything I might have missed. (There's a TODO to do so on the list jobs options class).
- I'm adding an inline note on a piece of code I changed that has nothing to do with labels.
- Marking this PR as not for submission until I surface projection and can write a snippet.

jskeet commented:
(Away at conferences at the moment, but will look when I can. I don't know whether filtering by label is even available, to be honest.)

amanda-tarafa commented:
No rush at all. And filtering by labels is not available, but we can at least make it possible on client side by surfacing projection.

amanda-tarafa commented:
@jskeet PTAL (no rush at all) thanks!
Added integration test and snippet for filtering listed jobs by labels.
Not sure about the place where I included the snippet at all but couldn't get to have either two snippets or even to see alsos per member.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/JobsTest.cs:
```diff
@@ -44,11 +46,200 @@ public void ListJobs()
 
             // Find all jobs started in the last minute.
-            var recentJobs = client.ListJobs().TakeWhile(job => job.Statistics.CreationTime >= jobToFind.Statistics.CreationTime).ToList();
+            var recentJobs = client.ListJobs().TakeWhile(job => job.Statistics.CreationTime >= oneMinuteAgo).ToList();
```
Hmm. I'm surprised the compiler didn't point out that I wasn't using the variable. Thanks for that!

And yes, we are relying on the ordering, but it's documented:

> The job list is sorted in reverse chronological order, by job creation time. 

It wouldn't hurt to add that to a comment though :)

amanda-tarafa replied:
Adding comment on the use of TakeWhile.

I think the compiler is not set to catch unused variables, for instance, in this same test the variable table wasn't being used either because the $ was missing from the front of the query string.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/JobsTest.cs:
```diff
@@ -67,5 +69,210 @@ public void ListJobs_FullProjection()
+            var jobToFind = client.CreateQueryJob($"SELECT * FROM {table}", null, new QueryOptions() { Labels = JobLabels});
+
+            // Find the job after listing with full projection.
+            var jobFound = (from job in client.ListJobs(new ListJobsOptions() { Projection = ProjectionEnum.Full })
```
I'd write this without the query expression:

```csharp
var options = new ListJobsOptions { Projection = ProjectionEnum.Full };
var jobFound = client.ListJobs(options).Single(job => job.Reference.JobId == jobToFind.Reference.JobId);
```

amanda-tarafa replied:
Done, definitely cleaner that way.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/JobsTest.cs:
```diff
@@ -67,5 +69,210 @@ public void ListJobs_FullProjection()
+        {
+            var client = BigQueryClient.Create(_fixture.ProjectId);
+            var table = client.GetTable(_fixture.ProjectId, _fixture.DatasetId, _fixture.HighScoreTableId);
+            var options = new QueryOptions() { Labels = JobLabels };
```
Just as a matter of style, I'd suggest omitting the `()` when calling a parameterless constructor with an object initializer. It doesn't make any difference here, but I think it's good to have a consistent style in the snippets, and the brevity is appealing :)

amanda-tarafa replied:
Done, here and elsewhere.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/JobsTest.cs:
```diff
@@ -67,5 +69,210 @@ public void ListJobs_FullProjection()
+            var client = BigQueryClient.Create(_fixture.ProjectId);
+
+            var table = client.GetTable(_fixture.ProjectId, _fixture.DatasetId, _fixture.HighScoreTableId);
+            var jobToFind = client.CreateQueryJob($"SELECT * FROM {table}", null, new QueryOptions() { Labels = JobLabels});
```
See later comment about parentheses for empty constructor argument lists.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/JobsTest.cs:
```diff
@@ -67,5 +69,210 @@ public void ListJobs_FullProjection()
+            VerifyJobLabels(jobFound?.Resource?.Configuration.Labels);
+        }
+
+        [Fact]
```
I've generally omitted integration tests if I have a snippet that tests the same thing. (Basically I regard snippets as integration tests in their own right.) I can't easily check right now, but do these tests check anything new? If so, they're very welcome. If they're effectively duplicating snippets, I think we're better off without them.

amanda-tarafa replied:
Removed integration tests that were duplicating snippets. Left 3 of them since they are going a little further in testing than the corresponding snippets are. I'm adding comments on this PR for the three tests that remained to signal what they are adding.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/JobsTest.cs:
```diff
@@ -67,5 +69,210 @@ public void ListJobs_FullProjection()
+            Assert.ThrowsAny<GoogleApiException>(() => client.CreateQueryJob($"SELECT * FROM {table}", null, options));
+        }
+
+        internal static IDictionary<string, string> JobLabels =>
```
Could you move this to the top of the file? (We tend to write fields/properties, constructors, then methods. Sometimes the properties are after the constructor, and we're not terribly consistent on that.)

amanda-tarafa replied:
Done.
At some point I though of adding that to the fixture (is used in the load test and the query test also) but wasn't sure about adding the VerifyJobLabels to the fixture as well

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/JobsTest.cs:
```diff
@@ -67,5 +69,210 @@ public void ListJobs_FullProjection()
+            {
+                // This one is a little absurd for jobs because labels can only be set
+                // on job creation. This means not to set the label because this is copying
+                // the bahaviour of labels for other resources like DataSets and Table.
```
Nit: bahaviour => behaviour (or behavior if we're going for US spelling; I don't mind either way)

amanda-tarafa replied:
Done

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.IntegrationTests/JobsTest.cs:
```diff
@@ -67,5 +69,210 @@ public void ListJobs_FullProjection()
+                // Testing for it anyways because the api allows is.
+                { "one_label", null },
+                { "another-label-2", "label_value_2" },
+                { "yet_another_label", string.Empty }
```
My personal preference is for "" over string.Empty, but I don't much mind either way (and I don't think it's worth worrying about :)

amanda-tarafa replied:
Changed to "" to try and keep everything the same, I don't have a conscious preference, I'm just used to string.Empty I think.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2.Snippets/JobCreationOptionsSnippets.cs:
```diff
@@ -0,0 +1,88 @@
+            string destinationUri = $"gs://{bucketName}/{objectName}";
+
+            // Just a couple examples of jobs marked with labels:
+            // (These jobs will most certainly be created somewhere else.)
```
Not sure what you mean by this parenthesized line.
(We haven't generally added snippets for options, but there's certainly no harm in doing so.)

amanda-tarafa replied:
With the parenthesized line I meant that you wouldn't create jobs "just for the sake of listing them", that is, that those jobs would have been created elsewhere as part of normal execution of the application.

My original intention was not to add the snippet to the options, I wanted to add it as a second sample for ListJobs(*) (list jobs and filter by labels) but couldn't find a way to do so. Don't know yet if it's possible to have two snippets per method signature, don't know if the second snippet is not generated at all or if it is just not included as part of the class docs, and if this is by design or not, wanted to ask you about it after looking at the code a little.

amanda-tarafa replied:
My original intention was not to add the snippet to the options, I wanted to add it as a second sample for ListJobs(*) (list jobs and filter by labels) but couldn't find a way to do so. Don't know yet if it's possible to have two snippets per method signature, don't know if the second snippet is not generated at all or if it is just not included as part of the class docs, and if this is by design or not, wanted to ask you about it after looking at the code a little.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/JobCreationOptions.cs:
```diff
@@ -44,5 +46,19 @@ public abstract class JobCreationOptions
+        /// You can use these to organize and group your jobs.
+        /// Label keys and values can be no longer than 63 characters,
+        /// can only contain lowercase letters, numeric characters,
+        /// underscores and dashes.International characters are allowed.
```
Nit: space after the period.

amanda-tarafa replied:
Done

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/JobCreationOptions.cs:
```diff
@@ -44,5 +46,19 @@ public abstract class JobCreationOptions
+        /// If a label value is <code>null</code>, it won't be added to the job's labels.
+        /// A label value can be <see cref="string.Empty"/>.
+        /// Label keys must start with a letter
+        /// and each label in the list must have a different key.
```
Not sure we need to mention about each label having to have a different key - given that it's a dictionary, that would only not be the case if someone provided a very odd comparer or their own dictionary implementation. (I think the fact that it *is* a dictionary is enough.)

amanda-tarafa replied:
Agreed, done.

amanda-tarafa commented:
@jskeet  Addressed all comments. Just a question regarding the snippet:

My first thought was to add it as a second snippet to `ListJobs(*)` but couldn't. If it's not really possible to have more than one snippet per method signature (it might be that I am missing something), is this by design or just for simplicity of the snippet handling code? Does it make sense for me to look into allowing more than one snippet?
(If you are OK with the snippet on the options, any answer to the above questions can be handled on a separate PR I think).

Thanks!

jskeet commented:
Will have to check when I'm on a laptop instead of a phone, but I'm sure it's fine :)

And yes, at the moment we only have support for one snippet per method. I think multiple snippets could end up being a bit overwhelming anyway, to be honest.
